python/test_infra/ttir_builder.py

```diff
@@ -0,0 +1,342 @@
+
+    # ----- Private helpers -----
+
+    @staticmethod
+    def _get_name(operand: Operand) -> str:
+        """Retrieves descriptive operand name."""
+        # Try to call get_name() if it exists, otherwise return operand.name.
+        return getattr(operand, "get_name", lambda: None)() or getattr(
+            operand, "name", None
+        )
+
+    def _store_golden(self, operand: Operand, golden: Golden) -> None:
+        """Maps `operand` to `golden`."""
+        assert (
+            self._goldens.get(operand) == None
+        ), f"Golden for {operand} already exists."
+        self._goldens[operand] = golden
+
+    def _get_golden(self, operand: Operand) -> Golden:
+        """
+        Retrieves stored golden for `operand`.
+
+        If no golden is found for `operand`, generates a random one.
+        """
+        golden = self._goldens.get(operand)
+
+        if golden is None:
+            golden = Golden.generate_random(self.get_shape(operand))
+            self._store_golden(operand, golden)
+
+        return golden
+
+    def _get_golden_tensor(self, operand: Operand) -> torch.Tensor:
+        return self._get_golden(operand).tensor
+
+    def _get_operand_constraint_attr(
+        self,
+        num_operands: int,
+        operand_constraints: Optional[List[tt.OperandConstraint]] = None,
+    ) -> tt.ir.OperandConstraintAttr:
+        """
+        Helper method to prepack operand constraints given as a list of enums to a list of
+        tt.ir.OperandConstraintAttr and wrap that list in an tt.ir.OperandConstraintAttr.
+
+        If no `operand_constraints` are passed, `tt.OperandConstraint.Any` will be used for each
+        operand.
+        """
+        operand_constraints = (
+            operand_constraints
+            if operand_constraints is not None
+            else [tt.OperandConstraint.Any for _ in range(num_operands)]
+        )
+
+        return tt.ir.OperandConstraintAttr.get(
+            self._ctx,
+            [
+                tt.ir.OperandConstraintAttr.get(self._ctx, operand_constraint)
+                for operand_constraint in operand_constraints
+            ],
+        )
+
+    @property
+    def _default_dtype(self) -> Type:
+        return F32Type.get(self._ctx)
+
+    def _get_type(self, input: Operand):
+        """
+        Helper method which retrieves underlying mlir Type of Operand based on which concrete
+        type it is.
+
+        We always expect it to be a RankedTensorType.
+        """
+        if isinstance(input, Value):
+            typ = input.type
+        elif isinstance(input, OpView):
+            typ = input.operation.result.type
+        elif isinstance(input, Operation):
+            typ = input.result.type
+        else:
+            raise TypeError(f"Invalid input {type(input)}")
+
+        assert isinstance(typ, RankedTensorType), "Only ranked tensors are supported"
+
+        return typ
+
+    # ----- Utility factories -----
+
+    def ranked_tensor(
```

Comment:
Maybe we should call it `ranked_tensor_type`, just to explicitly disambiguate types from ops.